kmath-tensors/src/commonMain/kotlin/space/kscience/kmath/tensors/core/algebras/DoubleLinearOpsTensorAlgebra.kt

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright 2018-2021 KMath contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package space.kscience.kmath.tensors.core.algebras
+
+import space.kscience.kmath.tensors.api.LinearOpsTensorAlgebra
+import space.kscience.kmath.nd.as1D
+import space.kscience.kmath.nd.as2D
+import space.kscience.kmath.tensors.api.TensorStructure
+import space.kscience.kmath.tensors.core.*
+import space.kscience.kmath.tensors.core.checkSquareMatrix
+import space.kscience.kmath.tensors.core.choleskyHelper
+import space.kscience.kmath.tensors.core.cleanSymHelper
+import space.kscience.kmath.tensors.core.luHelper
+import space.kscience.kmath.tensors.core.luMatrixDet
+import space.kscience.kmath.tensors.core.luMatrixInv
+import space.kscience.kmath.tensors.core.luPivotHelper
+import space.kscience.kmath.tensors.core.pivInit
+import kotlin.math.min
+
+
+public class DoubleLinearOpsTensorAlgebra :
+    LinearOpsTensorAlgebra<Double>,
+    DoubleTensorAlgebra() {
+
+    override fun TensorStructure<Double>.inv(): DoubleTensor = invLU(1e-9)
+
+    override fun TensorStructure<Double>.det(): DoubleTensor = detLU(1e-9)
+
+    public fun TensorStructure<Double>.luFactor(epsilon: Double): Pair<DoubleTensor, IntTensor> =
+        computeLU(tensor, epsilon) ?:
+        throw RuntimeException("Tensor contains matrices which are singular at precision $epsilon")
+
+    public fun TensorStructure<Double>.luFactor(): Pair<DoubleTensor, IntTensor> = luFactor(1e-9)
+
+    public fun luPivot(
+        luTensor: TensorStructure<Double>,
+        pivotsTensor: TensorStructure<Int>
+    ): Triple<DoubleTensor, DoubleTensor, DoubleTensor> {
+        checkSquareMatrix(luTensor.shape)
+        check(
+            luTensor.shape.dropLast(2).toIntArray() contentEquals pivotsTensor.shape.dropLast(1).toIntArray() ||
+                    luTensor.shape.last() == pivotsTensor.shape.last() - 1
+        ) { "Inappropriate shapes of input tensors" }
+
+        val n = luTensor.shape.last()
+        val pTensor = luTensor.zeroesLike()
+        for ((p, pivot) in pTensor.matrixSequence().zip(pivotsTensor.tensor.vectorSequence()))
+            pivInit(p.as2D(), pivot.as1D(), n)
+
+        val lTensor = luTensor.zeroesLike()
+        val uTensor = luTensor.zeroesLike()
+
+        for ((pairLU, lu) in lTensor.matrixSequence().zip(uTensor.matrixSequence())
```

Comment:
Rewrite as:
```
Tensor.matrixSequence()
            .zip(uTensor.matrixSequence())
            .zip(luTensor.tensor.matrixSequence())
            .forEach { (pairLU, lu) ->
                val (l, u) = pairLU
                luPivotHelper(l.as2D(), u.as2D(), lu.as2D(), n)
            }
```

Also, add a notice that this code will be able to be improved if [KT-13516](https://youtrack.jetbrains.com/v2/issue/KT-13516) is fixed.